File: lib/services/paymee_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:url_launcher/url_launcher.dart';

class PaymeeService {
  static const String apiKey = "98ce51ef202cfa8dc9bf31f3153face567f32ef5";
  static const int vendorId = 4101;

  // -----------------------------
  // CREATE PAYMENT
  // -----------------------------
  static Future<String> createPayment({
    required double amount, // in DT
    required String orderId,
    String firstName = "Test",
    String lastName = "Client",
    String email = "test@paymee.tn",
    String phone = "22222222",
    String returnUrl =
    "https://fulgorous-malakai-unpersonalized.ngrok-free.dev/payment/success",
    String cancelUrl =
    "https://fulgorous-malakai-unpersonalized.ngrok-free.dev/payment/cart",
    String webhookUrl =
    "https://fulgorous-malakai-unpersonalized.ngrok-free.dev/payment/webhook",
  }) async {
    final payload = {
      "vendor": vendorId,
      "amount": amount,
      "note": "Order $orderId",
      "first_name": firstName,
      "last_name": lastName,
      "email": email,
      "phone": phone,
      "return_url": returnUrl,
      "cancel_url": cancelUrl,
      "webhook_url": webhookUrl,
      "order_id": orderId,
    };

    print("Sending Paymee Payload: $payload");

    try {
      final response = await http.post(
        Uri.parse("https://sandbox.paymee.tn/api/v2/payments/create"),
        headers: {
          "Content-Type": "application/json",
          "Accept": "application/json",
          "Authorization": "Token $apiKey",
        },
        body: jsonEncode(payload),
      );

      final data = jsonDecode(response.body);
      print("Paymee Response: $data");

      if (data['status'] == true && data['data']?['payment_url'] != null) {
        return data['data']['payment_url'];
      } else {
        throw Exception(data['message'] ?? "Paymee failed");
      }
    } catch (e) {
      print("Paymee Network Error: $e");
      rethrow;
    }
  }

  // -----------------------------
  // OPEN PAYMENT PAGE
  // -----------------------------
  static Future<void> launchPayment(String url) async {
    if (!await launchUrl(Uri.parse(url),
        mode: LaunchMode.externalApplication)) {
      throw Exception("Could not open Paymee payment page");
    }
  }

  // -----------------------------
  // FULL FLOW
  // -----------------------------
  static Future<void> pay({
    required double amount,
    required String orderId,
  }) async {
    print("🔵 STEP 1: Creating Paymee order...");
    final url = await createPayment(amount: amount, orderId: orderId);

    print("🟡 STEP 2: Redirecting to Paymee...");
    await launchPayment(url);
  }
}

File: web/9arini/src/services/paymeeService.js
const API_KEY = "778a52b7cd6d034cbe74bf109d9ea9061f236962";
const VENDOR_ID = 4150;

export const createPayment = async (amount, orderId) => {
  const payload = {
    vendor: VENDOR_ID,
    
    // If your previous error said "Amount must be between 0.1 and 50000",
    // and you sent 300, and it failed... 
    // It's safer to send a standard float DT amount first.
    // But if you are sure about millimes, keep * 1000.
    // I will use standard float to be safe based on V2 docs.
    amount: parseFloat(amount), 

    note: `Order ${orderId}`,
    first_name: "Test",
    last_name: "Client",
    email: "test@paymee.tn",
    phone: "22222222",
    
    // CRITICAL FIX: MUST BE HTTPS
    return_url: "https://laraine-rightable-unpresumptively.ngrok-free.dev/payment/success", 
    cancel_url: "https://laraine-rightable-unpresumptively.ngrok-free.dev/payment/cart",
    webhook_url:"https://laraine-rightable-unpresumptively.ngrok-free.dev/payment/webhook",

    order_id: orderId
  };

  console.log("Sending Paymee Payload:", payload);

  try {
    const response = await fetch(
      "https://sandbox.paymee.tn/api/v2/payments/create",
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Accept": "application/json",
          "Authorization": `Token ${API_KEY}`
        },
        body: JSON.stringify(payload)
      }
    );

    const data = await response.json();
    console.log("Paymee Response:", data);

    if (data.status === true && data.data?.payment_url) {
      return data.data.payment_url;
    } else {
      console.error("Paymee Rejected:", data);
      throw new Error(data.message || "Paymee failed");
    }

  } catch (error) {
    console.error("Paymee Network Error:", error);
    throw error;
  }
};
